src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -2159,6 +2160,501 @@ class AutoDiffBatchNormTrainingRev
 struct WhileOpEnzymeOpsRemover
     : public EnzymeOpsRemoverOpInterface::ExternalModel<WhileOpEnzymeOpsRemover,
                                                         stablehlo::WhileOp> {
+private:
+#define DEBUG_TYPE "enzymexla-stablehlo-while-mincut"
+
+  // A node in the compute graph.
+  // Operation nodes have outgoing edges to value nodes that they produce and
+  // incoming nodes from values they take as operands.
+  struct Node {
+    Operation *O;
+    Value V;
+    enum Type {
+      NONE,
+      VAL,
+      OP,
+    } type;
+
+    Node(Operation *O) : O(O), type(OP){};
+    Node(Value V) : V(V), type(VAL){};
+    Node() : type(NONE){};
+
+    bool operator<(const Node N) const {
+      if (type != N.type)
+        return type < N.type;
+      else if (type == OP)
+        return O < N.O;
+      else if (type == VAL)
+        return V.getAsOpaquePointer() < N.V.getAsOpaquePointer();
+      else
+        return true;
+    }
+    void dump() const {
+      if (type == VAL)
+        llvm::errs() << "[" << V << ", "
+                     << "Value"
+                     << "]\n";
+      else if (type == OP)
+        llvm::errs() << "[" << *O << ", "
+                     << "Operation"
+                     << "]\n";
+      else
+        llvm::errs() << "["
+                     << "NULL"
+                     << ", "
+                     << "None"
+                     << "]\n";
+    }
+  };
+
+  typedef std::map<Node, std::set<Node>> Graph;
+
+  static void dump(Graph &G) {
+    for (auto &pair : G) {
+      pair.first.dump();
+      for (const auto &N : pair.second) {
+        llvm::errs() << "\t";
+        N.dump();
+      }
+    }
+  }
+
+  // parent is populated with a path from each connected leaf node of G to one
+  // of the Value in Source.
+  static inline void bfs(const Graph &G, const llvm::SetVector<Value> &Sources,
+                         std::map<Node, Node> &parent) {
+    std::deque<Node> q;
+    for (auto V : Sources) {
+      Node N(V);
+      parent.emplace(N, Node());
+      q.push_back(N);
+    }
+
+    // Standard BFS Loop
+    while (!q.empty()) {
+      auto u = q.front();
+      q.pop_front();
+      auto found = G.find(u);
+      if (found == G.end())
+        continue;
+      for (auto v : found->second) {
+        if (parent.find(v) == parent.end()) {
+          q.push_back(v);
+          parent.emplace(v, u);
+        }
+      }
+    }
+  }
+
+  // Whether or not an operation can be moved from the forward region to the
+  // reverse region or vice-versa.
+  static inline bool isMovable(Operation *op) {
+    return mlir::isPure(op) && op->getNumRegions() == 0;
```

Comment:
longer term we may want to turn this into an interface (e.g. because things like reduceop have regions, but should be fine)

Reply:
or more generally the issue is whether any op within a region uses a value defined outside the region

Reply:
Yes, the code only looks at operands to figure out if an operation can be moved: 

https://github.com/EnzymeAD/Enzyme-JAX/blob/ba778b6dfc540de2c638e0caeff222621727c2d0/src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp#L2352-L2357